www/content/essays/web-security-basics-with-htmx.md

```diff
@@ -0,0 +1,322 @@
+
+Inserting user-generated text directly into a script tag should never be necessary, but there *are* some situations where you might let users customize their CSS or customize HTML attributes. Handling those properly will be discussed down below.
+
+## Secure your cookies
+
+The best way to do authentication with htmx is using cookies. And because htmx encourages interactivity primarily through first-party HTML APIs, it is usually trivial to enable the browser's best cookie security features. These three in particular:
+
+* `Secure` - only send the cookie via HTTPS, never HTTP
+* `HttpOnly` - don't make the cookie available to JavaScript via `document.cookie`
+* `SameSite=Lax` - don't allow other sites to use your cookie to make requests, unless it's just a plain link
+
+To understand what these protect you against, let's go over the basics. If you come from JavaScript SPAs, where it's common to authenticate using the `Authorization` header, you might not be familiar with how cookies work. Fortunately they're very simple. (Please note: this is not an "authentication with htmx" tutorial, just an overview of cookie tokens generally)
+
+If your users log in with a `<form>`, their browser will send your server an HTTP request, and your server will send back a response that looks something like this:
+
+```http
+HTTP/2.0 200 OK
+Content-Type: text/html
+Set-Cookie: token=asd8234nsdfp982
+
+[HTML content]
+```
+
+That token corresponds to the user's current login session. From now on, every time that user makes a request to any route at `yourdomain.com`, the browser will include that cookie from `Set-Cookie` in the HTTP request.
+
+```http
+GET /users HTTP/1.1
+Host: yourdomain.com
+Cookie: token=asd8234nsdfp982
+```
+
+Each time someone makes a request to your server, it needs to parse out that token and determine if it's valid. Simple enough.
+
+You can also set options on that cookie, like the ones I recommended above. How to do this differs depending on the programming language, but the outcome is always an HTTP request that looks like this:
+
+```http
+HTTP/2.0 200 OK
+Content-Type: text/html
+Set-Cookie: token=asd8234nsdfp982; Secure; HttpOnly; SameSite=Lax
+
+[HTML content]
+```
+
+So what do the options do?
+
+The first one, `Secure`, ensures that the browser will not send the cookie over an insecure HTTP connection, only a secure HTTPS connection. Sensitive info, like a user's login token, should *never* be sent over an insecure connection.
+
+The second option, `HttpOnly`, means that the browser will not expose to the cookie to JavaScript, ever (i.e. it won't be in [`document.cookie`](https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie)). Even if someone is able to insert a malicious script, like in the `evilwebsite.com` example above, that malicious script cannot access the user's cookie or send it to `evilwebsite.com`. The browser will only attach the cookie when the request is made to the website the cookie came from.
```

Comment:
`s/the browser will not expose to the cookie/the browser will not expose the cookie/`